From faabba5a4c78151dd89a51ad1c2b43017d91537a Mon Sep 17 00:00:00 2001
From: Jeffy Chen <jeffy.chen@rock-chips.com>
Date: Tue, 12 May 2020 15:35:40 +0800
Subject: [PATCH 06/98] input: Fix focus loss when reattaching keyboard

Launch weston-terminal, start typing, then detach and reattach the
keyboard. Focus is lost.

Save focus with notify_keyboard_focus_out() when releasing the last
keyboard, and restore it when attaching a new one.

Signed-off-by: Jeffy Chen <jeffy.chen@rock-chips.com>
---
 libweston/input.c         | 17 +++++++++++------
 libweston/libinput-seat.c |  4 ++++
 2 files changed, 15 insertions(+), 6 deletions(-)

diff --git a/libweston/input.c b/libweston/input.c
index e51faf78..02c91e66 100644
--- a/libweston/input.c
+++ b/libweston/input.c
@@ -2806,9 +2806,11 @@ notify_keyboard_focus_out(struct weston_seat *seat)
 		weston_pointer_cancel_grab(pointer);
 
 	if (focus) {
+		if (seat->saved_kbd_focus)
+			wl_list_remove(&seat->saved_kbd_focus_listener.link);
+
 		seat->use_saved_kbd_focus = true;
 		seat->saved_kbd_focus = focus;
-		assert(seat->saved_kbd_focus_listener.notify == NULL);
 		seat->saved_kbd_focus_listener.notify =
 			destroy_device_saved_kbd_focus;
 		wl_signal_add(&focus->destroy_signal,
@@ -4163,13 +4165,16 @@ weston_keyboard_reset_state(struct weston_keyboard *keyboard)
 WL_EXPORT void
 weston_seat_release_keyboard(struct weston_seat *seat)
 {
-	seat->keyboard_device_count--;
-	assert(seat->keyboard_device_count >= 0);
-	if (seat->keyboard_device_count == 0) {
-		weston_keyboard_set_focus(seat->keyboard_state, NULL);
-		weston_keyboard_cancel_grab(seat->keyboard_state);
+	assert(seat->keyboard_device_count > 0);
+
+	if (seat->keyboard_device_count == 1) {
+		notify_keyboard_focus_out(seat);
+		seat->keyboard_device_count = 0;
+
 		weston_keyboard_reset_state(seat->keyboard_state);
 		seat_send_updated_caps(seat);
+	} else {
+		seat->keyboard_device_count--;
 	}
 }
 
diff --git a/libweston/libinput-seat.c b/libweston/libinput-seat.c
index 189b96d5..58478026 100644
--- a/libweston/libinput-seat.c
+++ b/libweston/libinput-seat.c
@@ -145,6 +145,10 @@ device_added(struct udev_input *input, struct libinput_device *libinput_device)
 			seat->led_update(seat, seat->keyboard_state->xkb_state.leds);
 	}
 
+	if (device->seat_caps & EVDEV_SEAT_KEYBOARD &&
+	    seat->keyboard_device_count == 1)
+		evdev_notify_keyboard_focus(seat, &udev_seat->devices_list);
+
 	return 0;
 }
 
-- 
2.39.5

